File: GDTMobApp/Release/GDTAds/GDTUnifiedNativeAd/ViewController/UnifiedNativeAdPortraitFeedViewController.m
//
//  UnifiedNativeAdPortraitVideoViewController.m
//  GDTMobApp
//
//  Created by royqpwang on 2019/5/16.
//  Copyright © 2019 Tencent. All rights reserved.
//

#import "UnifiedNativeAdPortraitFeedViewController.h"
#import "GDTUnifiedNativeAd.h"
#import "UnifiedNativePortraitCollectionViewCell.h"
#import "GDTAppDelegate.h"
#import "UnifiedNativeAdPortraitDetailViewController.h"

@interface UnifiedNativeAdPortraitFeedViewController () <GDTUnifiedNativeAdDelegate, UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout>

@property (nonatomic, strong) GDTUnifiedNativeAd *nativeAd;
@property (nonatomic, strong) UICollectionView *collectionView;
@property (nonatomic, strong) NSMutableArray *adArray;

@end

@implementation UnifiedNativeAdPortraitFeedViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self initView];
    self.adArray = [NSMutableArray new];
    
    //---- 视频配置项，整段注释可使用外部VC开关控制
//    self.videoConfig.videoMuted = NO;
//    self.videoConfig.autoPlayPolicy = GDTVideoAutoPlayPolicyNever; // 手动控制播放
//    self.videoConfig.userControlEnable = YES;
//    self.videoConfig.autoResumeEnable = NO;
//    self.videoConfig.detailPageEnable = NO;
    //-----
    
    if (self.useToken) {
        self.nativeAd = [[GDTUnifiedNativeAd alloc] initWithPlacementId:self.placementId token:self.token];
    } else {
        self.nativeAd = [[GDTUnifiedNativeAd alloc] initWithPlacementId:self.placementId];
    }
    self.nativeAd.delegate = self;
    self.nativeAd.minVideoDuration = self.minVideoDuration;
    self.nativeAd.maxVideoDuration = self.maxVideoDuration;
    [self.nativeAd loadAdWithAdCount:10];
}

- (void)initView
{
    [self.view addSubview:self.collectionView];
    [self.collectionView registerClass:[UnifiedNativePortraitCollectionViewCell class]
            forCellWithReuseIdentifier:@"UnifiedNativePortraitCollectionViewCell"];
    [self.collectionView registerClass:[UICollectionViewCell class] forCellWithReuseIdentifier:@"UICollectionViewCell"];
}

#pragma mark - GDTUnifiedNativeAdDelegate
- (void)gdt_unifiedNativeAdLoaded:(NSArray<GDTUnifiedNativeAdDataObject *> *)unifiedNativeAdDataObjects error:(NSError *)error
{
    if (error) {
        NSLog(@"error %@", error);
        return;
    }
    
    for (GDTUnifiedNativeAdDataObject *dataObject in unifiedNativeAdDataObjects) {
        if (dataObject.isVideoAd) {
            [self.adArray addObject:@"test"];
            [self.adArray addObject:@"test"];
            [self.adArray addObject:dataObject];
        }
        NSLog(@"extraInfo: %@", dataObject.extraInfo);
        NSLog(@"eCPM:%ld eCPMLevel:%@", [dataObject eCPM], [dataObject eCPMLevel]);
    }
    [self.collectionView reloadData];
}

#pragma mark - UICollectionViewDataSource

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section
{
    return self.adArray.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath
{
    id item = self.adArray[indexPath.row];
    if ([item isKindOfClass:[GDTUnifiedNativeAdDataObject class]]) {
        UnifiedNativePortraitCollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"UnifiedNativePortraitCollectionViewCell" forIndexPath:indexPath];
        [cell setupWithUnifiedNativeAdDataObject:item];
        return cell;
        
    } else {
        UICollectionViewCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"UICollectionViewCell" forIndexPath:indexPath];
        cell.backgroundColor = [UIColor colorWithRed:((10 * indexPath.row) / 255.0) green:((20 * indexPath.row)/255.0) blue:((30 * indexPath.row)/255.0) alpha:1.0f];
        return cell;
    }
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    return CGSizeMake(self.view.bounds.size.width / 2 - 20, 320);
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath
{
    id item = self.adArray[indexPath.row];
    if ([item isKindOfClass:[GDTUnifiedNativeAdDataObject class]]) {
        UnifiedNativeAdPortraitDetailViewController *vc = [[UnifiedNativeAdPortraitDetailViewController alloc] init];
        vc.dataObject = item;
        vc.dataObject.videoConfig = self.videoConfig;
        [self.navigationController pushViewController:vc animated:YES];
    }
}

#pragma mark - property getter
- (UICollectionView *)collectionView
{
    if (!_collectionView) {
        UICollectionViewFlowLayout *layout = [[UICollectionViewFlowLayout alloc] init];
        [layout setScrollDirection:UICollectionViewScrollDirectionVertical];
        layout.minimumLineSpacing = 2;
        _collectionView.frame = self.view.bounds;
        _collectionView.translatesAutoresizingMaskIntoConstraints = YES;
        _collectionView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
        _collectionView = [[UICollectionView alloc] initWithFrame:self.view.bounds collectionViewLayout:layout];
        _collectionView.backgroundColor = [UIColor grayColor];
        _collectionView.delegate = self;
        _collectionView.dataSource = self;
        _collectionView.accessibilityIdentifier = @"collectionView_id";
    }
    return _collectionView;
}


@end

File: GDTMobSample-Swift/GDTAdTypes/UnifiedNativeAdViewController/ViewController/UnifiedNativeAdPortraitFeedViewController.swift
//
//  UnifiedNativeAdPortraitFeedViewController.swift
//  GDTMobSample-Swift
//
//  Created by 胡城阳 on 2019/10/28.
//  Copyright © 2019 Tencent. All rights reserved.
//

import UIKit

class UnifiedNativeAdPortraitFeedViewController: UnifiedNativeAdBaseViewController,GDTUnifiedNativeAdDelegate,UICollectionViewDelegate,UICollectionViewDataSource,UICollectionViewDelegateFlowLayout {
    

    var nativeAd:GDTUnifiedNativeAd?
    var collectionView:UICollectionView?
    var adArray:Array<Any> = []
    override func viewDidLoad() {
        super.viewDidLoad()
        setCollectionView()
        initView()
        self.adArray = Array()
        
        //---- 视频配置项，整段注释可使用外部VC开关控制
        self.placementId = "3050349752532954"; // 需要只出视频的广告位，请联系商务同学开通
        self.videoConfig!.videoMuted = false;
        self.videoConfig!.autoPlayPolicy = GDTVideoAutoPlayPolicy.never; // 手动控制播放
        self.videoConfig!.userControlEnable = true;
        self.videoConfig!.autoResumeEnable = false;
        self.videoConfig!.detailPageEnable = false;
        //-----
        self.nativeAd = GDTUnifiedNativeAd.init(appId: self.appId, placementId: self.placementId)
        self.nativeAd?.delegate = self
        self.nativeAd?.maxVideoDuration = self.maxVideoDuration
        self.nativeAd?.loadAd(withAdCount: 10)
        // Do any additional setup after loading the view.
    }
    
    func initView(){
        self.view.addSubview(self.collectionView!)
        self.collectionView!.register(UnifiedNativePortraitCollectionViewCell.self, forCellWithReuseIdentifier: "UnifiedNativePortraitCollectionViewCell")
        self.collectionView!.register(UICollectionViewCell.self, forCellWithReuseIdentifier: "UICollectionViewCell")
    }
    
    func setCollectionView(){
        let layout:UICollectionViewFlowLayout = UICollectionViewFlowLayout()
        layout.scrollDirection = .vertical
        collectionView = UICollectionView(frame: self.view.bounds, collectionViewLayout: layout)
        collectionView!.translatesAutoresizingMaskIntoConstraints = true
        collectionView!.autoresizingMask = UIViewAutoresizing(rawValue: UIViewAutoresizing.flexibleWidth.rawValue | UIViewAutoresizing.flexibleHeight.rawValue)
        collectionView!.backgroundColor = UIColor.gray
        collectionView!.delegate = self
        collectionView!.dataSource = self
    }
}
// GDTUnifiedNativeAdDelegate
extension UnifiedNativeAdPortraitFeedViewController{
    func gdt_unifiedNativeAdLoaded(_ unifiedNativeAdDataObjects: [GDTUnifiedNativeAdDataObject]?, error: Error?) {
        if (error != nil) {
            print("error is \(error)")
            return
        }
        
        for dataObject in unifiedNativeAdDataObjects! {
            if dataObject.isVideoAd{
                self.adArray.append("test")
                self.adArray.append("test")
                self.adArray.append(dataObject)
            }
            
        }
        self.collectionView!.reloadData()
    }
}

extension UnifiedNativeAdPortraitFeedViewController{
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return self.adArray.count
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let item = self.adArray[indexPath.row]
        if item is GDTUnifiedNativeAdDataObject {
            let cell:UnifiedNativePortraitCollectionViewCell = collectionView.dequeueReusableCell(withReuseIdentifier: "UnifiedNativePortraitCollectionViewCell", for: indexPath) as! UnifiedNativePortraitCollectionViewCell
            cell.setupWithUnifiedNativeAdDataObject(nativeAdDataObject: item as! GDTUnifiedNativeAdDataObject)
            return cell
        }else{
            let cell:UICollectionViewCell = collectionView.dequeueReusableCell(withReuseIdentifier: "UICollectionViewCell", for: indexPath)
            cell.backgroundColor = UIColor.red
            return cell
        }
    }
    
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return CGSize(width: self.view.bounds.width / 2 - 20 , height: 320)
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let item = self.adArray[indexPath.row]
        if item is GDTUnifiedNativeAdDataObject {
            let vc:UnifiedNativeAdPortraitDetailViewController = UnifiedNativeAdPortraitDetailViewController()
            vc.dataObject = (item as! GDTUnifiedNativeAdDataObject)
            vc.dataObject?.videoConfig = self.videoConfig
            self.navigationController?.pushViewController(vc, animated: true)
        }
    }
}
